staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,633 @@
+// in the the FinishPacket function
+func (qs *QueueSetImpl) addQueues(n int) {
+	for i := 0; i < n; i++ {
+		qs.queues = append(qs.queues, &Queue{
+			Packets: []*Packet{},
+		})
+	}
+}
+
+// ===========================================================================
+// ===========================================================================
+
+// Quiesce controls whether this system is quiescing.  Passing a
+// non-nil handler means the system should become quiescent, a nil
+// handler means the system should become non-quiescent.  A call
+// to Wait while the system is quiescent will be rebuffed by
+// returning `quiescent=true`.  If all the queues have no requests
+// waiting nor executing while the system is quiescent then the
+// handler will eventually be called with no locks held (even if
+// the system becomes non-quiescent between the triggering state
+// and the required call).
+//
+// The filter uses this for a priority level that has become
+// undesired, setting a handler that will cause the priority level
+// to eventually be removed from the filter if the filter still
+// wants that.  If the filter later changes its mind and wants to
+// preserve the priority level then the filter can use this to
+// cancel the handler registration.
+func (qs *QueueSetImpl) Quiesce(eh EmptyHandler) {
+	qs.lock.Lock()
+	defer qs.lock.Unlock()
+	if eh == nil {
+		qs.quiescent = false
+		return
+	}
+	// Here we check whether there are any requests queued or executing and
+	// if not then fork an invocation of the EmptyHandler.
+	if qs.numPacketsEnqueued == 0 && qs.GetRequestsExecuting() == 0 {
+		// fork an invocation of the EmptyHandler.
+		go func() {
+			eh.HandleEmpty()
+		}()
+	}
+	qs.quiescent = true
+}
+
+// Wait in the happy case, shuffle shards the given request into
+// a queue and eventually dispatches the request from that queue.
+// Dispatching means to return with `quiescent==false` and
+// `execute==true`.  In one unhappy case the request is
+// immediately rebuffed with `quiescent==true` (which tells the
+// filter that there has been a timing splinter and the filter
+// re-calcuates the priority level to use); in all other cases
+// `quiescent` will be returned `false` (even if the system is
+// quiescent by then).  In the non-quiescent unhappy cases the
+// request is eventually rejected, which means to return with
+// `execute=false`.  In the happy case the caller is required to
+// invoke the returned `afterExecution` after the request is done
+// executing.  The hash value and hand size are used to do the
+// shuffle sharding.
+func (qs *QueueSetImpl) Wait(hashValue uint64, handSize int32) (quiescent, execute bool, afterExecution func()) {
+	// TODO(aaron-prindle) verify what should/shouldn't be locked!!!!
+	// TODO(aaron-prindle) collapse all of FQ into one layer/lock (vs 3)
+	//   currently able to collapse to 1 impl layer and 2 locks...
+
+	qs.LockAndSyncTime()
```

Comment:
Since there is no blanket `defer qs.lock.Unlock()`, you have to be sure to unlock before every return.

Reply:
updated